pkg/cloudprovider/providers/gce/gce_loadbalancer.go

```diff
@@ -875,6 +937,44 @@ func (gce *GCECloud) firewallNeedsUpdate(name, serviceName, region, ipAddress st
 	return true, false, nil
 }
 
+func (gce *GCECloud) ensureHttpHealthCheckFirewall(region string, hosts []*gceInstance, hcName string, hcPort int32, isNodesHealthCheck bool) error {
+	// Prepare the firewall params for creating / checking.
+	desc := "Firewall for health checks from load-balancer."
+	sourceRanges := lbSrcRngsFlag.ipn
+	ports := []v1.ServicePort{{Protocol: "tcp", Port: hcPort}}
+
+	clusterId, err := gce.ClusterId.GetId()
+	if err != nil {
+		return fmt.Errorf("error getting cluster ID: %v", err)
+	}
+	fwNameSuffix := MakeHealthCheckFirewallNameSuffix(clusterId, hcName, isNodesHealthCheck)
+	fwName := makeFirewallName(fwNameSuffix)
+	fw, err := gce.service.Firewalls.Get(gce.projectID, fwName).Do()
+	if err != nil {
+		if !isHTTPErrorCode(err, http.StatusNotFound) {
+			return fmt.Errorf("error getting firewall for health checks: %v", err)
+		}
+		glog.Infof("Creating firewall %v for health checks.", fwName)
+		if err := gce.createFirewall(fwNameSuffix, region, desc, sourceRanges, ports, hosts); err != nil {
+			return err
+		}
+		glog.Infof("Created firewall %v for health checks.", fwName)
+		return nil
+	}
+	// Validate firewall fields.
+	if fw.Description != desc ||
+		!sameStringArray(fw.SourceRanges, sourceRanges.StringSlice()) ||
+		!sameStringArray(packProtocolsPortsFromFirewallAlloweds(fw.Allowed), packProtocolsPortsFromServicePorts(ports)) {
```

Comment:
I think these helper funcs (`packProtocolsPortsFromFirewallAlloweds` and `packProtocolsPortsFromServicePorts`) are nice, but overkill and may cause an error eventually.  You're making them easy to compare; however, they are actually different. ServicePorts can contain multiple ports, one with TCP and one with UDP.  The load balancer is assembled using the first protocol, so the other ports' protocols don't matter.  The `firewallObject` func performs this logic, but you're not doing that on the other side of the equals.  It's currently working because the health check only requires one port with one protocol.

Since these helpers are only used here, I'd suggest going with simpler/smaller logic that matches current implementation in gce_loadbalancer. I also used the same logic in ILB.
https://github.com/nicksardo/kubernetes/blob/a6fe71bd6c7746ddd21fa008420d1c7ddac69698/pkg/cloudprovider/providers/gce/gce_loadbalancer_internal.go#L68 
Thoughts?